gui/job-details.lua

```diff
@@ -0,0 +1,425 @@
+            )
+            return
+        end
+    end
+
+    guimat.MaterialDialog{
+        prompt = 'Please select a new material for input '..idx,
+        none_caption = 'any material',
+        mat_filter = function(mat,parent,mat_type,mat_index)
+            return dfhack.job.isSuitableMaterial(obj.iobj, mat_type, mat_index, obj.iobj.item_type)
+        end,
+        on_select = self:callback('setMaterial', obj)
+    }:show()
+end
+
+function JobDetails:canChangeTrait()
+    local idx, obj = self.list:getSelected()
+    return obj ~= nil and not is_caste_mat(obj.iobj)
+end
+
+function JobDetails:onChangeTrait()
+    local idx, obj = self.list:getSelected()
+    guimat.ItemTraitsDialog{
+        job_item = obj.iobj,
+        prompt = 'Please select traits for input '..idx,
+        none_caption = 'no traits',
+    }:show()
+end
+
+local function ScrJobDetails()
+    return df.global.game.main_interface.job_details
+end
+
+local function show_job_details()
+    local scr = ScrJobDetails()
+    if not scr.open -- dfhack.gui.matchFocusString('dwarfmode/JobDetails')
+    then
+        qerror("This script needs to be run from a job details screen")
+    end
+
+    local job
+    if scr.context == df.job_details_context_type.BUILDING_TASK_LIST then
+        job = scr.jb
+    elseif scr.context == df.job_details_context_type.MANAGER_WORK_ORDER then
+        job = scr.wq
+    end
+    if job == nil then
+        qerror("Unhandled screen context: ".. df.job_details_context_type[scr.context])
+    end
+
+    JobDetailsScreen{ job = job, context = scr.context }:show()
+end
+
+-- --------------------
+-- DetailsHotkeyOverlay
+--
+
+local LABEL_TEXT = 'Configure job inputs'
+local LABEL_TEXT_LENGTH = string.len( LABEL_TEXT )
+
+DetailsHotkeyOverlay = defclass(DetailsHotkeyOverlay, overlay.OverlayWidget)
+DetailsHotkeyOverlay.ATTRS{
+    default_pos={x=0,y=0},
+    default_enabled=true,
+    viewscreens="override this in a subclass",
+    frame={w= 1   -- [
+            + 6   -- Ctrl+d
+            + 2   -- :_
+            + LABEL_TEXT_LENGTH -- LABEL_TEXT
+            + 1   -- ]
+         , h= 1
+        },
+}
+
+function DetailsHotkeyOverlay:init()
+    self:addviews{
+        widgets.TextButton{
+            view_id = 'button',
+            frame={t=0, l=0, r=0, h=1},
+            label=LABEL_TEXT,
+            key='CUSTOM_CTRL_D',
+            on_activate=show_job_details,
+        },
+    }
+end
+
+DetailsHotkeyOverlay_ManagerWorkOrder = defclass(DetailsHotkeyOverlay_ManagerWorkOrder, DetailsHotkeyOverlay)
+DetailsHotkeyOverlay_ManagerWorkOrder.ATTRS{
+    default_pos={x=5, y=5}, -- {x=5, y=5} is right above the job title
+    viewscreens='dwarfmode/JobDetails/MANAGER_WORK_ORDER',
+}
+
+DetailsHotkeyOverlay_BuildingTask = defclass(DetailsHotkeyOverlay_BuildingTask, DetailsHotkeyOverlay)
+DetailsHotkeyOverlay_BuildingTask.ATTRS{
+    default_pos={x=-120, y=6}, -- {x=-120, y=6} is right above the job title on all but smallest widths
```

Comment:
I'd move this to the right by 5 so it still looks good at the smallest width

Reply:
I "reserved" the whole line using `frame={w=1000, h= 1}`, so that now with `dfhack.screen.getWindowSize()` I can move the hotkey around and it always fits.

It would've been nice to be able to move the widget itself. IIRC that wasn't possible last time I checked (half a year ago? because of label placement in the workorder-recheck.lua).

Reply:
I improved upon this. The logic is as follows now.

0. Fix the distance from the right edge; make the widget slightly wider than necessary to allow moving the text around
1. determine the calculated x1 coordinate of the frame, which is the `screen width` minus the `distance from the right` minus the `widget width`
2. if the x1 < 0, set it to 0 --- making it the real position of the widget (I *think* this works better than built-in methods for some reason, although it's a few experiments back)
3. is the real left edge position of the widget too far to the left (`x1 < 6`), move the text to the right (`offset = 6 - x1`).

This works well enough until you try to narrow the game window to the minimum and then widen it again. Then the `self.frame.r` is changed from `102` to `77` and stays that way. That's why the 4th step is needed

4. restore original `self.frame.r`

Even with this after widening the window it doesn't immediately get the right position, but another small resizing in any direction sets it right.